packages/react/src/components/Drawer/index.tsx

```diff
@@ -0,0 +1,179 @@
+  portal?: React.RefObject<HTMLElement> | HTMLElement;
+}
+
+const Drawer = forwardRef<HTMLDivElement, DrawerProps>(
+  (
+    {
+      children,
+      className,
+      position,
+      open = false,
+      hideScrim = false,
+      focusTrap = false,
+      focusOptions = {},
+      portal,
+      onClose,
+      style,
+      ...props
+    },
+    ref
+  ) => {
+    const drawerRef = useSharedRef(ref);
+    const openRef = useRef(!!open);
+    const previousActiveElementRef = useRef<HTMLElement>(
+      null
+    ) as React.MutableRefObject<HTMLElement | null>;
+    const { initialFocus: focusInitial, returnFocus: focusReturn } =
+      focusOptions;
+    const [isTransitioning, setIsTransitioning] = useState(!!open);
+
+    const handleClose = useCallback(() => {
+      // istanbul ignore next
+      if (open && typeof onClose === 'function') {
+        onClose();
+      }
+    }, [open, onClose]);
+
+    useEffect(() => {
+      // jsdom does not trigger transitionend event
+      // istanbul ignore next
+      const transitionEndHandler = () => setIsTransitioning(false);
+
+      document.addEventListener('transitionend', transitionEndHandler);
+      return () => {
+        document.removeEventListener('transitionend', transitionEndHandler);
+      };
+    }, [setIsTransitioning]);
+
+    useEffect(() => {
+      if (openRef.current !== open) {
+        setIsTransitioning(true);
+      }
+
+      openRef.current = open;
+    }, [open, setIsTransitioning]);
+
+    useLayoutEffect(() => {
+      if (open) {
+        previousActiveElementRef.current =
+          document.activeElement as HTMLElement;
+
+        const initialFocusElement = resolveElement(focusInitial);
+        if (initialFocusElement) {
+          initialFocusElement.focus();
+        } else {
+          const focusable = drawerRef.current?.querySelector(
+            focusableSelector
+          ) as HTMLElement;
+          if (focusable) {
+            focusable.focus();
+          } else {
+            // fallback focus
+            drawerRef.current?.focus();
+          }
+        }
+      } else if (previousActiveElementRef.current) {
+        const returnFocusElement = resolveElement(focusReturn);
+        if (returnFocusElement) {
+          returnFocusElement.focus();
+        } else {
+          // fallback focus
+          previousActiveElementRef.current?.focus();
+        }
+      }
+    }, [open, focusInitial, focusReturn]);
+
+    useEscapeKey(
+      { callback: handleClose, active: open, defaultPrevented: true },
+      [onClose]
+    );
+
+    // istanbul ignore next
+    if (!isBrowser()) {
+      return null;
+    }
+
+    const portalElement = resolveElement(portal);
+
+    return createPortal(
+      <>
+        <ClickOutsideListener
+          onClickOutside={handleClose}
+          mouseEvent={open ? undefined : false}
+          touchEvent={open ? undefined : false}
+          target={drawerRef}
+        >
+          <FocusTrap
```

Comment:
I don't know what the right answer is, but when focus is trapped, the drawer seems to act like a modal. Similar to modals, should we make the content outside the drawer hidden to screen reader users when focus is trapped?

Reply:
On a related note, when focus is not trapped, I wonder how valid it is to have the scrim visible. It seems like you should only have the scrim visible when focus is trapped. Or, tabbing outside of the drawer could collapse the drawer, which is another pattern I have seen

Reply:
> should we make the content outside the drawer hidden to screen reader users when focus is trapped

Yes, good catch. Oversight on my part I will address this.

Reply:
> On a related note, when focus is not trapped, I wonder how valid it is to have the scrim visible. It seems like you should only have the scrim visible when focus is trapped. Or, tabbing outside of the drawer could collapse the drawer, which is another pattern I have seen

There's a lot of nuance to this where I'm not 100% sure what the behavior should be. 

The [apg dialog pattern](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/) calls this out as a best practice:

> Inert content outside an active dialog is typically visually obscured or dimmed so it is difficult to discern, and in some implementations, attempts to interact with the inert content cause the dialog to close.

I was trying to have a declarative nature for both focus trap and the scrim, but maybe that's a mistake. @schne324 do you have any thoughts on this?

Reply:
I agree with the APG best practice here. Does UX have a strong opinion to NOT scrim stuff out when the drawer is open?

Reply:
@schne324 There will be future implementations of the drawer that are not focus traps (e.g. it's a slide out drawer that does not cover content) that are more like non-modal dialogs OR push content out of the way so it needs to be configurable.

So it sounds like where we may want to land is that focus trap / scrim should not be handled independently but controlled via a single prop. Does that make sense?

Reply:
ah yes, I agree. Focus trap and scrim should be tied at the hip.

Reply:
Updated this to capture the drawer as behavior `modal` or `non-modal`.